File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        # 1. If both p and q are smaller than root, LCA is in the left subtree
        if p.val < root.val and q.val < root.val:
            return self.lowestCommonAncestor(root.left, p, q)
        
        # 2. If both p and q are greater than root, LCA is in the right subtree
        elif p.val > root.val and q.val > root.val:
            return self.lowestCommonAncestor(root.right, p, q)
        
        # 3. If we are here, one is on the left and one is on the right 
        # (or one equals the root), so THIS node is the LCA.
        else:
            return root

File: Data Structures & Algorithms/same-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode* left;
 *     TreeNode* right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(
 *         int x,
 *         TreeNode* left,
 *         TreeNode* right
 *     ) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        // Both nodes are empty, so they match.
        if (p == nullptr && q == nullptr) {
            return true;
        }

        // Only one node is empty, so the trees differ.
        if (p == nullptr || q == nullptr) {
            return false;
        }

        // Current values and both subtrees must match.
        return p->val == q->val &&
               isSameTree(p->left, q->left) &&
               isSameTree(p->right, q->right);
    }
};